Add symmetric tensor tutorial to docs as appendix (#316)
Jutho/TensorKit.jl: docs/src/appendix/symmetric_tutorial.md

irreps $a$ and $b$. These fusion rules are called *Abelian* if the tensor product of any two
irreps corresponds to exactly one irrep. We will return to the implications of *non-Abelian*
fusion rules [later](@ref ss_non_abelian). For the case of the $\mathbb{Z}_2$ irreps, the fusion
rules are Abelian, and are given by addition modulo 2,
```math
0 \otimes 0 \cong 0, \quad 0 \otimes 1 \cong 1, \quad 1 \otimes 0 \cong 1, \quad 1 \otimes 1 \cong 0.
```
To see how these fusion rules arise, we can consider the action of the symmetry
transformation $XX$ on the possible two-site basis states, each of which corresponds to a
tensor product of representations. We can see that $XX$ acts trivially on both $\ket{+}
\otimes \ket{+}$ and $\ket{-} \otimes \ket{-}$ meaning these transform under the trivial
representation, which gives the first and last entries of the fusion rules. Similarly, $XX$
acts with a minus sign on both $\ket{+} \otimes \ket{-}$ and $\ket{-} \otimes \ket{+}$,
meaning these transform under the sign representation, which gives the second and third
entries of the fusion rules. Having introduced this notion of 'fusing' irreps, we can now

> associate associate a well-defined *coupled irrep* to each of the four two-site basis

Reviewer: ```suggestion
associate a well-defined *coupled irrep* to each of the four two-site basis
```

Just as in the $\mathbb{Z}_2$ case, it is obvious that we cannot directly construct the
creation and annihilation operators as instances of a `TensorMap(..., V ← V)` since they are
not invariant under conjugation by the symmetry operator. However, it is possible to
construct them as `TensorMap`s using an *auxiliary vector space*, based on the following
intuition. The creation operator $a^+$ violates particle number conservation by mapping the
occupation number $n$ to $n + 1$. From the point of view of representation theory, this
process can be thought of as the *fusion* of an `U1Irrep(n)` with an `U1Irrep(1)`, naturally
giving the fusion product `U1Irrep(n + 1)`. This means we can represent $a^+$ as a
`TensorMap(..., V ← V ⊗ A)`, where the auxiliary vector space `A` contains the $+1$ irrep
with degeneracy 1, `A = U1Space(1 => 1)`. Similarly, the decrease in occupation number when
acting with $a^-$ can be thought of as the *splitting* of an `U1Irrep(n)` into an
`U1Irrep(n - 1)` and an `U1Irrep(1)`, leading to a representation in terms of a
`TensorMap(..., A ⊗ V ← V)`. Based on these observations, we can represent the matrix

> elements {eq}`eq:bosonopmatel` as blocks labeled by the $\mathrm{U}(1)$ fusion trees

Reviewer: equation reference

Each irrep can be labeled as $l = D(p,q)$ where $p$ and $q$ are refered to as the *Dynkin
labels*. The eigenvalue of the quadratic Casimir for a given irrep is given by
[Freudenthal's formula](https://en.wikipedia.org/wiki/Weyl_character_formula#Freudenthal's_formula),
```math
\Omega(D(p,q)) = \frac{1}{3} (p^2 + q^2 + 3p + 3q + pq).
```
Using SUNRepresentations.jl, we can compute the Casimir as
```@example symmetric_tutorial
function casimir(l::SU3Irrep)
    p, q = dynkin_label(l)
    return (p^2 + q^2 + 3 * p + 3 * q + p * q) / 3
end
```
If we use the adjoint representation of $\mathrm{SU}(3)$ as physical space, the Heisenberg
exchange interaction can then be constructed as
```@example symmetric_tutorial
V = Vect[SUNIrrep{3}](SU3Irrep("8") => 1)
TT = zeros(ComplexF64, V ⊗ V ← V ⊗ V)
for (s, f) in fusiontrees(TT)
    l3 = f.uncoupled[1]
    l4 = f.uncoupled[2]
    k = f.coupled
    TT[s, f] .= (casimir(k) - casimir(l3) - casimir(l4)) / 2
end
subblocks(TT)
```
Circling back to our earlier remark, we clearly see that the fusion trees of this tensor
indeed have non-trivial vertex labels.
```@example symmetric_tutorial

> f = collect(fusiontrees(TT))[3][2]

Author: Huh, that's strange, the fusion trees must be in a different order than when I ran the original tutorial. Is the fusion tree order deterministic given a specific homspace, and could it be that this order changed between March 2024 and now?

    which case we basically provide a narrated walk-through of the corresponding code.

```@example symmetric_tutorial
using LinearAlgebra
using TensorKit
using WignerSymbols
using SUNRepresentations
using Test # for showcase testing
```

## Level 0: The 1-dimensional Ising model

As the most basic example, we will consider the
[1-dimensional transverse-field Ising model](https://en.wikipedia.org/wiki/Transverse-field_Ising_model),
whose Hamiltonian is given by

```math
\begin{equation}
\label{eq:isingham}
H = -J \left (\sum_{\langle i, j \rangle} Z_i Z_j + g \sum_{i} X_i\right).
\end{equation}
```

Here, $X_i$ and $Z_i$ are the
[Pauli operators](https://en.wikipedia.org/wiki/Pauli_matrices) acting on site $i$, and the
first sum runs over pairs of nearest neighbors $\langle i, j \rangle$. This model has a
global $\mathbb{Z}_2$ symmetry, as it is invariant under the transformation $U H U^\dagger =
H$ where the symmetry transformation $U$ is given by a global spin flip,

```math
\begin{equation}
\label{eq:z2gen}
U = \prod_i X_i.
\end{equation}
```

We will circle back to the implications of this symmetry later.

> As a warmup we will implement the Hamiltonian \eqref{eq:isingham} in the standard way by

Author: I tried to figure this out, but with no success. It should be possible to configure this in Mathjax somehow (see e.g. [here](https://stackoverflow.com/questions/65020117/mathjax-2-7-7-eqref-color)). However, I find it very unclear how to actually configure the Mathjax enginge for Documenter.jl in detail, and some trial and error attempts lead me nowhere.

\ket{-} \otimes \ket{+} &\mapsto \ket{+} \otimes \ket{-}, \nonumber \\
\ket{-} \otimes \ket{-} &\mapsto \ket{+} \otimes \ket{+}. \nonumber \\
\end{align}
```

We will denote the trivial $\mathbb{Z}_2$ irrep by $'0'$, corresponding to a local $\ket{+}$
state, and the sign irrep by $'1'$, corresponding to a local $\ket{-}$ state. Given this
identification, we can naturally associate the tensor product of basis vectors in the irrep
basis to the tensor product of the corresponding $\mathbb{Z}_2$ irreps. One of the key
questions of the [representation theory of groups](representation_theory) is how the tensor
product of two irreps can be decomposed into a direct sum of irreps. This decomposition is
encoded in what are often called the
[*fusion rules*](https://en.wikipedia.org/wiki/Fusion_rules),
```math
a \otimes b \cong \bigoplus_c N_c^{ab} c,
```
where $N_{ab}^c$ encodes the number of times the irrep $c$ occurs in the tensor product of
irreps $a$ and $b$. These fusion rules are called *Abelian* if the tensor product of any two
irreps corresponds to exactly one irrep. We will return to the implications of *non-Abelian*

> fusion rules [later](@ref ss_non_abelian). For the case of the $\mathbb{Z}_2$ irreps, the fusion

Reviewer 2: In the `FusionStyle` trait, we are not actually using the "abelian" vs "non-abelian" nomenclature, because at some point it became confusing if you also want to consider Vec[G]. For a non-abelian group, the "fusion" of the simple objects of Vec[G] is unique. However, it does depend on the order: `a ⊗ b != b ⊗ a`. Hence, there cannot be a braiding. On the other hand, for `Rep[G]`, it does hold that `a ⊗ b == b ⊗ a`, also for a non-abelian group.

So then, what does abelian mean: does it mean unique fusion, or independent of the order fusion? 

Note that I have no problem with calling the group abelian or non-abelian; but I would prefer to avoid calling fusion rules abelian.

inserted into the appropriate matrix block. So, for the purpose of this tutorial **we will
interpret a symmetric `TensorMap` simply as a list of fusion trees, to each of which
corresponds a certain reduced matrix element**. In TensorKit.jl, these reduced tensor
elements corresponding to the fusion trees of a `TensorMap` can be accessed through the
[`subblocks`](@ref) method.

!!! note
    In general, such a reduced matrix element is not necessarily a scalar, but rather an array
    whose size is determined by the degeneracy of the irreps in the codomain and domain of the
    fusion tree. For this reason, a reduced matrix element associated to a given fusion tree is
    also referred to as an *array block*. In the following we will use terms 'reduced matrix
    element', 'array block' or just 'block' interchangeably. However, it should be remembered
    that these are distinct from the matrix blocks in the block-diagonal decomposition of the
    tensor.

### [Fusion trees and how to use them](@id sss_fusion_trees)

This view of the underlying symmetry structure in terms of fusion trees and corresponding

> array blocks is a very convenient way of working with the `TensorMap` type. Consider a

Reviewer 2: Would there be a reason not to use:

```suggestion
subblocks is a very convenient way of working with the `TensorMap` type. Consider a
```

$\mathbb{Z}_2$ irreps are self-dual and have Abelian fusion rules. We will come back to
these other properties when discussion more involved applications. Given some `TensorMap`,
the method `TensorKit.fusiontrees(t::TensorMap)` returns an iterator over all pairs of
splitting and fusion trees that label the blocks of `t`.

We can now put this into practice by directly constructing the $ZZ$ operator in the irrep
basis as a $\mathbb{Z}_2$-symmetric `TensorMap`. We will do this in three steps:

- First we construct the physical space at each site as a $\mathbb{Z}_2$-graded vector space.
- Then we initialize an empty `TensorMap` with the correct domain and codomain vector spaces built from the previously constructed physical space.
- And finally we iterate over all splitting and fusion tree pairs and manually fill in the corresponding nonzero blocks of the operator.

After the basis transform to the irrep basis, we can view the two-dimensional complex
physical vector space we started with as being spanned by the trivial and sign irrep of
$\mathbb{Z}_2$. In the language of TensorKit.jl, this can be implemented as a `Z2Space`, an
alias for a

> $\mathbb{Z}_2$-[graded vector space](@ref GradedSpace)

Reviewer 2: I would simply call this graded space here. Explicitly mentioning Z_2-graded might be confusing, because it is this way only because `Irrep[Z{2}]` is isomorphic to Z_2, but it is `Irrep[Z{2}]` that is providing the grading.

```suggestion
[graded vector space](@ref GradedSpace)
```

As the most basic example, we will consider the
[1-dimensional transverse-field Ising model](https://en.wikipedia.org/wiki/Transverse-field_Ising_model),
whose Hamiltonian is given by

```math
\begin{equation}
\label{eq:isingham}
H = -J \left (\sum_{\langle i, j \rangle} Z_i Z_j + g \sum_{i} X_i\right).
\end{equation}
```

Here, $X_i$ and $Z_i$ are the
[Pauli operators](https://en.wikipedia.org/wiki/Pauli_matrices) acting on site $i$, and the
first sum runs over pairs of nearest neighbors $\langle i, j \rangle$. This model has a
global $\mathbb{Z}_2$ symmetry, as it is invariant under the transformation $U H U^\dagger =
H$ where the symmetry transformation $U$ is given by a global spin flip,

```math
\begin{equation}
\label{eq:z2gen}
U = \prod_i X_i.
\end{equation}
```

We will circle back to the implications of this symmetry later.

As a warmup we will implement the Hamiltonian \eqref{eq:isingham} in the standard way by
encoding the matrix elements of the single-site operators $X$ and $Z$ into an array of
complex numbers, and then combine them in a suitable way to get the Hamiltonian terms.
Instead of using plain Julia arrays, we will use a representation in terms of `TensorMap`s

> over complex vector spaces. These will essentially just be wrappers around base arrays at

Reviewer 2: ```suggestion
over complex vector spaces. These are essentially just wrappers around base arrays at
```

this point, but their construction requires some consideration of the notion of *spaces*,
which generalize the notion of `size` for arrays. Each of the operators $X$ and $Z$ acts on

> a local 2-dimensional complex vector space. In the context of TensorKit.jl such a space can

Reviewer 2: ```suggestion
a local 2-dimensional complex vector space. In the context of TensorKit.jl, such a space can
```

questions of the [representation theory of groups](representation_theory) is how the tensor
product of two irreps can be decomposed into a direct sum of irreps. This decomposition is
encoded in what are often called the
[*fusion rules*](https://en.wikipedia.org/wiki/Fusion_rules),
```math
a \otimes b \cong \bigoplus_c N_c^{ab} c,
```
where $N_{ab}^c$ encodes the number of times the irrep $c$ occurs in the tensor product of
irreps $a$ and $b$. These fusion rules are called *Abelian* if the tensor product of any two
irreps corresponds to exactly one irrep. We will return to the implications of irreps with *non-Abelian* fusion rules [later](@ref ss_non_abelian). For the case of the $\mathbb{Z}_2$
irreps, the fusion rules are Abelian, and are given by addition modulo 2,
```math
0 \otimes 0 \cong 0, \quad 0 \otimes 1 \cong 1, \quad 1 \otimes 0 \cong 1, \quad 1 \otimes 1 \cong 0.
```
To see how these fusion rules arise, we can consider the action of the symmetry
transformation $XX$ on the possible two-site basis states, each of which corresponds to a
tensor product of representations. We can see that $XX$ acts trivially on both $\ket{+}

> \otimes \ket{+}$ and $\ket{-} \otimes \ket{-}$ meaning these transform under the trivial

Reviewer 2: ```suggestion
\otimes \ket{+}$ and $\ket{-} \otimes \ket{-}$, meaning these transform under the trivial
```

```

which can be used to label a subblock of a `TensorMap` corresponding to a two-site operator.
This object should actually be seen as a *pair of fusion trees*. The first member of the
pair, related to the codomain of the `TensorMap`, is referred to as the *splitting tree* and
encodes how the *coupled charge* $c$ splits into the *uncoupled charges* $s_1$ and $s_2$.
The second member of the pair, related to the domain of the `TensorMap`, is referred to as
the *fusion tree* and encodes how the uncoupled charges $f_1$ and $f_2$ fuse to the coupled
charge $c$. Both the splitting and fusion tree can be represented as a [`FusionTree`](@ref)
instance. You will find such a `FusionTree` has the following properties encoded into its
fields:

- `uncoupled::NTuple{N,I}`: a list of `N` uncoupled charges of type `I<:Sector`
- `coupled::I`: a single coupled charge of type `I<:Sector`
- `isdual::NTuple{N,Bool}`: a list of booleans indicating whether the corresponding uncoupled charge is dual
- `innerlines::NTuple{M,I}`: a list of inner lines of type `I<:Sector` of length `M = N - 2`

> - `vertices::NTuple{L,T}`: list of fusion vertex labels of type `T` and length `L = N - 1`

Reviewer 2: The terminology "charge" is also used below in `flip_charge`, so probably a small comment here would be sufficient but useful.

flips the irreps of the uncoupled charges in the domain with respect to the codomain, as
shown in the diagrams above. Flipping a given `Z2Irrep` in the codomain can be implemented
by fusing them with the sign irrep `Z2Irrep(1)`, giving:

```@example symmetric_tutorial
flip_charge(charge::Z2Irrep) = only(charge ⊗ Z2Irrep(1))
for (s, f) in fusiontrees(ZZ)
    if s.uncoupled == map(flip_charge, f.uncoupled)
        ZZ[s, f] .= 1
    end
end
subblocks(ZZ)
```

Indeed, the resulting `TensorMap` exactly encodes the matrix elements of the $ZZ$ operator
shown in [the diagrams above](Z2_fusiontrees). The $X$ operator can be constructed in a
similar way. Since it is by definition diagonal in the irrep basis with matrix blocks
directly corresponding to the trivial and sign irrep, its construction is particularly
simple:

```@example symmetric_tutorial
X = zeros(ComplexF64, V ← V)
for (s, f) in fusiontrees(X)
    if only(f.uncoupled) == Z2Irrep(0)
        X[s, f] .= 1
    else
        X[s, f] .= -1
    end
end
subblocks(X)
```

Given these local operators, we can use them to construct the full manifestly
$\mathbb{Z}_2$-symmetric Hamiltonian.

!!! note

>     An important observation is that when explicitly imposing the $\mathbb{Z}_2$ symmetry we

Reviewer 2: ```suggestion
    An important observation is that, when explicitly imposing the $\mathbb{Z}_2$ symmetry, we
```

    directly constructed the full $ZZ$ operator as a single symmetric tensor. This in contrast
    to the case without symmetries, where we constructed a single-site $Z$ operator and then
    combined them into a two-site operator. Clearly this can no longer be done when imposing
    $\mathbb{Z}_2$, since a single $Z$ is not invariant under conjugation with the symmetry
    operator $X$. One might wonder whether it is still possible to construct a two-site
    Hamiltonian term by combining local objects. This is possible if one introduces an auxiliary
    index on the local tensors that carries a non-trivial charge. The intuition behind this will
    become more clear in the next example.

## Level 2: The $\mathrm{U}(1)$ Bose-Hubbard model

> For our next example we will consider the

Reviewer 2: ```suggestion
For our next example, we consider the
```

\end{equation}
```
This Hamiltonian is defined on the [Fock space](https://en.wikipedia.org/wiki/Fock_space) associated to a chain of bosons,
where the action bosonic creation, annihilation and number operators $a^+$, $a^-$ and $N =
a^+ a^-$ in the local occupation number basis is given by
```math
\begin{align}
\label{eq:bosonopmatel}
a^+ \ket{n} &= \sqrt{n + 1} \ket{n + 1} \\
a^- \ket{n} &= \sqrt{n} \ket{n - 1} \nonumber \\
N \ket{n} &= n \ket{n} \nonumber
\end{align}
```
Their bosonic nature can be summarized by the familiar the commutation relations
```math
\begin{align*}
\left[a_i^-, a_j^-\right] &= \left[a_i^+, a_j^+\right] = 0 \\
\left[a_i^-, a_j^+\right] &= \delta_{ij} \\
\left[N, a^+\right] &= a^+ \\
\left[N, a^-\right] &= -a^- \\
\end{align*}
```

This Hamiltonian is invariant under conjugation by the global particle number operator, $U H
U^\dagger = H$, where
```math
U = \sum_i N_i
```
This invariance corresponds to a $\mathrm{U}(1)$ particle number symmetry, which can again
be manifestly imposed when constructing the Hamiltonian terms as `TensorMap`s. From the

> representation theory of $\mathrm{U}(1)$ we know that it's irreps are all one-dimensional

Reviewer 2: ```suggestion
representation theory of $\mathrm{U}(1)$, we know that its irreps are all one-dimensional
```

the physical space. This will now be a $\mathrm{U}(1)$ graded vector space `U1Space`, where
each basis state $\ket{n}$ in the occupation number basis is represented by the
corresponding $\mathrm{U}(1)$ irrep `U1Irrep(n)` with degeneracy 1. While this physical
space is in principle infinite dimensional, we will impose a cutoff in occupation number at
a maximum of 5 bosons per site, giving a 6-dimensional vector space:

```@example symmetric_tutorial
cutoff = 5
V = U1Space(n => 1 for n in 0:cutoff)
```

We can now initialize the $a^+ a^-$, $a^- a^+$ and $N$ operators as empty `TensorMap`s with
the correct domain and codomain vector spaces, and fill in the nonzero subblocks associated
to [the fusion trees shown above](U1_fusiontrees). To do this we need access to the integer
label of the $\mathrm{U}(1)$ irreps in the fusion and splitting trees, which can be accessed
through the `charge` field of the `U1Irrep` type.

```@example symmetric_tutorial
a⁺a⁻ = zeros(ComplexF64, V ⊗ V ← V ⊗ V)
for (s, f) in fusiontrees(a⁺a⁻)
    if s.uncoupled[1] == only(f.uncoupled[1] ⊗ U1Irrep(1)) && s.uncoupled[2] == only(f.uncoupled[2] ⊗ U1Irrep(-1))

>         a⁺a⁻[s, f] .= sqrt(s.uncoupled[1].charge * f.uncoupled[2].charge)

Reviewer 3: Maybe use the `charge` function here, and all other instances below?

elements \eqref{eq:bosonopmatel} as subblocks labeled by the $\mathrm{U}(1)$ fusion trees

```@raw html
<center><img src="../img/symmetric_tutorial/bosonops.svg" alt="bosonops" class="color-invertible" style="zoom: 170%"/></center>
```

We can then combine these operators to get the appropriate Hamiltonian terms,

```@raw html
<center><img src="../img/symmetric_tutorial/bosonham.svg" alt="bosonham" class="color-invertible" style="zoom: 170%"/></center>
```

!!! note
    Although we have made a suggestive distinction between the 'left' and 'right' versions of
    the operators $a_L^\pm$ and $a_R^\pm$, one can actually be obtained from the other by
    permuting the physical and auxiliary indices of the corresponding `TensorMap`s. This
    permutation has no effect on the actual subblocks of the tensors due to the bosonic
    [`BraidingStyle`](@ref) of $\mathrm{U}(1)$ irreps, so the left and right operators can
    in essence be seen as the 'same' tensors. This is no longer the case when considering
    fermionic systems, where permuting indices can in fact change the subblocks as we will
    see next. As a consequence, it is much less clear how to construct two-site symmetric

>     operators in terms of local symmetric objects.

Author: I added a `FusionStyle` box, mentioning `UniqueFusion` and `MultipleFusion` and how they relate to the nomenclature commonly used in anyon models that I adopted here.

end
a⁺
```

```@example symmetric_tutorial
a⁻ = zeros(ComplexF64, A ⊗ V ← V)
for (s, f) in fusiontrees(a⁻)
    a⁻[s, f] .= sqrt(f.uncoupled[1].charge)
end
a⁻
```

It is then simple to check that this is indeed what we expect.
```@example symmetric_tutorial
@tensor a⁺a⁻_bis[-1 -2; -3 -4] := a⁺[-1; -3 1] * a⁻[1 -2; -4]
@tensor a⁻a⁺_bis[-1 -2; -3 -4] := a⁻[1 -1; -3] * a⁺[-2; -4 1]
@tensor N_bis[-1 ; -2] := a⁺[-1; 1 2] * a⁻[2 1; -2]

@test a⁺a⁻_bis ≈ a⁺a⁻ atol=1e-14
@test a⁻a⁺_bis ≈ a⁻a⁺ atol=1e-14
@test N_bis ≈ N atol=1e-14
```

!!! note
    From the construction of the Hamiltonian operators
    [in terms of creation and annihilation operators](bosonham) we clearly see that they are
    invariant under a transformation $a^\pm \to e^{\pm i\theta} a^\pm$. More generally, any
    invertible transformation on the auxiliary space leaves the resulting contraction unchanged.
    This ambiguity in the definition clearly shows that one should really always think in terms
    of the fully symmetric procucts of $a^+$ and $a^-$ rather than in terms of these operators
    themselves. In particular, one can always decompose such a symmetric product into the

>     [form above](bosonham) by means of an SVD.

Author: Good point, that wasn't clear at all. I rewrote it to clearly describe what kind of transformation I mean.

## Level 3: Fermions and the Kitaev model

While we have already covered quite a lot of ground towards understanding symmetric tensors
in terms of fusion trees and corresponding subblocks, the symmetries considered so far have
been quite 'simple' in the sense that sectors corresponding to irreps of $\mathbb{Z}_2$ and
$\mathrm{U}(1)$ have [*Abelian fusion rules*](@ref FusionStyle) and
[*bosonic exchange statistics*](@ref BraidingStyle).

> This means that the fusion of two irreps always gives a single irrep as the fusion product,

Reviewer 2: Is this too suggestive?
```suggestion
This means that the fusion of two irreps always gives a unique irrep as the fusion product,
```

essential to the fermionic nature of the system. Indeed, if we for example work out the
matrix element of $c_1^- c_2^+$ we find
```math
\begin{align*}
c_1^- c_2^+ \ket{1, 0} = c_1^- c_2^+ c_1^+ \ket{0, 0} = - c_2^+ c_1^- c_1^+ \ket{0, 0} = - c_2^+ (\mathbb{1} - c_1^+ c_1^-) \ket{0, 0} = - c_2^+ \ket{0, 0} = - \ket{0, 1}. \\
\end{align*}
```

Once we have these matrix elements the hard part is done, and we can straightforwardly
associate these to the following $f\mathbb{Z}_2$ fusion trees with corresponding reduced
tensor elements,

```@raw html
<center><img src="../img/symmetric_tutorial/fZ2_fusiontrees.svg" alt="fZ2_fusiontrees" class="color-invertible" style="zoom: 170%"/></center>
```

Given this information, we can go through the same procedure again to construct $c^+ c^-$,
$c^- c^+$ and $N$ operators as `TensorMap`s over $f\mathbb{Z}_2$-graded vector spaces.

```@example symmetric_tutorial
V = Vect[FermionParity](0 => 1, 1 => 1)
```

```@example symmetric_tutorial
c⁺c⁻ = zeros(ComplexF64, V ⊗ V ← V ⊗ V)
odd = FermionParity(1)
for (s, f) in fusiontrees(c⁺c⁻)
    if s.uncoupled[1] == odd && f.uncoupled[2] == odd && f.coupled == odd
        c⁺c⁻[s, f] .= 1
    end
end

> c⁺c⁻

Reviewer 2: Here and below, whatever we decide regarding the printing of `TensorMap`s (I agree with the general sentiment that the current printing output is not optimal), you probably intended to show the subblocks?
```suggestion
subblocks(c⁺c⁻)
```

representation spaces in more detail.

[From the representation theory of groups](https://en.wikipedia.org/wiki/Tensor_product_of_representations#Clebsch%E2%80%93Gordan_theory),
it is known that the product of two irreps can in turn be decomposed into a direct sum of
irreps, $V^{(l_1)} \otimes V^{(l_2)} \cong \bigoplus_{k} V^{(k)}$. The precise nature of
this decomposition, also refered to as the *Clebsch-Gordan problem*, is given by the
so-called *Clebsch-Gordan coefficients*, which we will denote as $C^{k}_{l_1,l_2}$. This set
of coefficients, which can be interpreted as a $\text{dim}\left( V^{(l_1)} \right) \times
\text{dim}\left( V^{(l_2)} \right) \times \text{dim}\left( V^{(k)} \right)$ array, that
encodes how a basis state $\ket{k,n} \in V^{(k)}$ corresponding to some term in the direct
sum can be decomposed into a linear combination of basis vectors $\ket{l_1,m_1} \otimes
\ket{l_2,m_2}$ of the tensor product space:
```math
\begin{equation}
\label{eq:cg_decomposition}
\ket{k,n} = \sum_{m_1, m_2} \left( C^{k}_{l_1,l_2} \right)^{n}_{m_1, m_2} \ket{l_1,m_1} \otimes \ket{l_2,m_2}.
\end{equation}

> ```

Reviewer 2: In this general discussion, it is kind of assumed that every irrep only appears once in the fusion output (`SimpleFusion`), or otherwise it is implied that `k` plays a different role than `l_1` and `l_2`, with the latter two really being irrep labels, while `k` a way to enumerate all irreps appearing in the fusion product. Maybe a `!!!note` box can be a solution?

This procedure works for any group symmetry, and all we need are matrix elements of the
operator in the irrep basis and the Clebsch-Gordan coefficients. In the following we
demonstrate this explicit procedure for the particular example of $G = \mathrm{SU}(2)$.
However, it should be noted that for general groups the Clebsch-Gordan coefficients may not
be as easy to compute (in general, no closed formulas exist). In addition, the procedure for
manually projecting out the reduced tensor elements requires being particularly careful
about the correspondence between the basis states used to define the original matrix
elements and those implied by the Clebsch-Gordan coefficients. Finally, for some symmetries
supported in TensorKit.jl, there are simply no Clebsch-Gordan coefficients. Therefore, it is
often easier and sometimes simply necessary to directly construct the symmetric tensor and
then fill in its reduced tensor elements based on some representation theory. Wel will cover
some examples of this below.

Having introduced and demonstrated the Clebsch-Gordan decomposition, the corresponding
coefficients and their role in symmetric tensors for the example of $\mathrm{SU}(2)$ using

> the WignerSymbols.jl package, we'll continue our discussion using only TensorKit.jl

Reviewer 2: ```suggestion
the WignerSymbols.jl package, we now continue our discussion using only TensorKit.jl
```

```@example symmetric_tutorial
get_reduced_element(SU2Irrep(1))
```
```@example symmetric_tutorial
get_reduced_element(SU2Irrep(2))
```
we can read off the entries
```math
\renewcommand\thickspace{\kern .01ex}
\left[ (\vec{S}_i \cdot \vec{S}_j)_\text{red} \right] \,\,\!\!
\begin{smallmatrix}
    1,1\\
    0\\
    1,1
\end{smallmatrix} = -2, \quad
\left[ (\vec{S}_i \cdot \vec{S}_j)_\text{red} \right] \,\,\!\!
\begin{smallmatrix}
    1,1\\
    1\\
    1,1
\end{smallmatrix} = -1, \quad
\left[ (\vec{S}_i \cdot \vec{S}_j)_\text{red} \right] \,\,\!\!
\begin{smallmatrix}
    1,1\\
    2\\
    1,1
\end{smallmatrix} = 1, \quad
```
These can then be used to construct the symmetric `TensorMap` representing the exchange
interaction:
```@example symmetric_tutorial
V = SU2Space(1 => 1)
SS = zeros(ComplexF64, V ⊗ V ← V ⊗ V)
for (s, f) in fusiontrees(SS)
    k = f.coupled
    SS[s, f] .= get_reduced_element(k)
end
subblocks(SS)
```

We demonstrated this entire procedure of extracting the reduced tensor elements of a
symmetric tensor map for each fusion tree by projecting out the corresponding fusion tensors
as an explicit illustration of how symmetric tensor maps work under the hood. In practice

> however, there's no need to perform this procedure explicitly. Given a dense array

Reviewer 2: ```suggestion
however, there is no need to perform this procedure explicitly. Given a dense array
```

'symmetric'](@ref ss_symmetries). In going through these examples we aim to provide a
relatively gentle introduction to the meaning of [symmetry sectors](@ref ss_sectors) and
[vector spaces](@ref ss_rep) within the context of TensorKit.jl, [how to initialize a
`TensorMap` over a given vector space](@ref ss_tensor_construction) and finally how to
manually set the data of a [symmetric `TensorMap`](@ref ss_tutorial_symmetries). We will
keep our discussion as intuitive and simple as possible, only adding as many technical
details as strictly necessary to understand each example. When considering a different
physical system of interest, you should then be able to adapt these recipes and the
intuition behind them to your specific problem at hand.

!!! note
    Many of these examples are readily implemented in the
    [TensorKitTensors.jl package](https://github.com/QuantumKitHub/TensorKitTensors.jl), in
    which case we basically provide a narrated walk-through of the corresponding code.

```@example symmetric_tutorial
using LinearAlgebra
using TensorKit
using WignerSymbols
using SUNRepresentations
using Test # for showcase testing
```

> ## Level 0: The 1-dimensional Ising model

Reviewer: Do we have to add "1-dimensional" here? Since we are always just looking at the building blocks, I don't think that actually makes a difference, so we might as well write "transverse-field Ising model" instead

  corresponding nonzero subblocks of the operator.

In TensorKit.jl, the representations of $\mathbb{Z}_2$ are represented as instances of the
[`Z2Irrep <: Sector`](@ref ZNIrrep) type. There are two such instances, corresponding to the
trivial irrep `Z2Irrep(0)` and the sign irrep `Z2Irrep(1)`. We can fuse irreps with the `⊗`
(`\otimes`) operator, which can for example be used to check their fusion rules,
```@example symmetric_tutorial
for a in values(Z2Irrep), b in values(Z2Irrep)
    println("$a ⊗ $b: $(a ⊗ b)")
end
```
After the basis transform to the irrep basis, we can view the two-dimensional complex
physical vector space we started with as being spanned by the trivial and sign irrep of
$\mathbb{Z}_2$. In the language of TensorKit.jl, this can be implemented as a `Z2Space`, an
alias for a [graded vector space](@ref GradedSpace) `Vect[Z2Irrep]`. To construct such a
graded space we have to specify which irreps it contains, and indicate the degenaracy of
each irrep. Here, our physical vector space contains the trivial irrep `Z2Irrep(0)` with
degeneracy 1 and the sign irrep `Z2Irrep(1)` with degeneracy 1. This means can define this

> space in the following way, and check its dimension:

Author: Introducing 'degeneracy' a bit more gently is definitely a good idea, but I don't want to introduce notation that I then don't use again, so I'll add a slightly rephrased version.

```@example symmetric_tutorial
V = Z2Space(0 => 1, 1 => 1)

> dim(V)

Reviewer 2: Maybe add the method to obtain the degeneracy here as well
```suggestion
dim(V)
dim(V, Z2Irrep(0)) == 1
dim(V, Z2Irrep(1))
```
